hbase-client/src/main/java/org/apache/hadoop/hbase/client/RpcConnectionRegistryCreator.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.client;
+
+import java.io.IOException;
+import java.net.URI;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.security.User;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Connection registry creator implementation for creating {@link RpcConnectionRegistry}.
+ */
+@InterfaceAudience.Private
+public class RpcConnectionRegistryCreator implements ConnectionRegistryCreator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RpcConnectionRegistryCreator.class);
+
+  @Override
+  public ConnectionRegistry create(URI uri, Configuration conf, User user) throws IOException {
+    assert protocol().equals(uri.getScheme());
+    LOG.debug("connect to hbase cluster with rpc bootstrap servers='{}'", uri.getAuthority());
+    Configuration c = new Configuration(conf);
+    c.set(RpcConnectionRegistry.BOOTSTRAP_NODES, uri.getAuthority());
+    return new RpcConnectionRegistry(c, user);
+  }
+
+  @Override
+  public String protocol() {
+    return "hbase+rpc";
```

Comment:
This part gets messy. I think the convention that has arisen around URI's with a '+' in their scheme section is used to indicate a protocol+transport. What does that mean for us? What do we even call our current RPC implementation of "Hadoop RPC plus protobuf cell back-channel", just "HBase RPC", so `hbaserpc`, which maybe is short for `hbaserpc+tcp`?  Say down the road we support an hbase rpc over HTTP or over GRPC (which itself supports http/2 and http/3 as transports, as well as grpc-web) or over UDP, what then? What do we call our existing Thrift ("hbase+thrift" ?) and REST gateways ("rest+http(s)" ?)

By contrast, the zookeeper client isn't a protocol at all. It's just a location of an expected service type. So then we can call it just "zookeeper" or "zk" for short.

I'm not saying we expect to have all these transport mechanisms, but we should think through what we want this part of our public API to look like and give precise meaning to the scheme section of the URI.

Reply:
I just followed the similar way with what we have in phoenix.

https://phoenix.apache.org/classpath_and_url.html

Agree that we should take care of the schema part. Maybe we could start a discussion thread again on the mailing list?

Reply:
IMO having the first part be just hbase to denote that your talking to hbase makes the most sense. Then the second part can be whatever. Based on what technology we have today and for years past and the foreseeable future, zk and rpc make sense. 

The suffix tells our code whether to parse the authority as a list of zk servers or as a list of bootstrap nodes. That unrelated to grpc vs custom rpc, etc. 

If someone wanted to create a grpc or http2 or w/e protocol, they'd still be talking to hbase and they'd still need someway to bootstrap the connection. So I bet the scheme would stay the same and which communication protocol to use would be a query parameter. 

We could argue what's most to spec but since the +suffix stuff is largely not spec'd I think we should consider user intuitiveness and how our code uses it.

Reply:
I suppose, the way to think about this connection URI is that is it just for locating and initiating a connection to a cluster. Even if we were to pursue a more complete specification as I've advocated, a client still has non-trivial reliance on other configuration parameters. It's unrealistic to expect that we could roll up those details in the scheme portion of the URI. You've convinced me.

Given the menu of options that we have available today, what you have implemented here seems fine. Can we add "hbase" without the "+..." part and let that be the default bootstrap mechanism for the current hbase version? In 2.6 that would be an alias for hbase+zk, for 3.0 that would be an alias for hbase+rpc.

Reply:
I think the scheme part should contain the 'zk' or 'rpc' information, so we could know how to decode the other part of the URI.

And for other parameters, we could add them as the parameter of the URI in the future.

Reply:
Or maybe another choice is to introduce a parameter like 'registry'. So the URI could be
```
hbase://zk1:2181,zk2:2181,zk3:2181/xxx?registry=zk
```

or
```
hbase://rs1:16010/?registry=rpc
```

Reply:
Personally I don't think that option works well because the point of the scheme is to tell the code how to parse the authority. For rpc vs zk, we need different parsing. It seems odd (even if possible) to put that in the query param. To me we could put things in the query param that affect anything else about the connection other than how to parse the uri

Reply:
IUC the standard Hbase client libraries don't support any other procol than HBaseRPC.
IF they were to ever support say REST and/or Thrift, then we could still add new variant for them as needed.
However, the current API probably has a lot of built-in assumptions of using RPC, and even if that were to be solved, achieving and maintaining parity for the alternate protocols would be monumental job, and we can make up new protocol variants for those when needed.

Reply:
One other issue we run into with Phoenix quite bit is that the cluster configurations still have to match in several aspects, like timeouts, TLS/SASL settings, etc, otherwise the client either can not even connect, or experiences errors due timeout / buffer size etc mismathces.

I think that some of that may also be a problem when configuring replication.

Reply:
> One other issue we run into with Phoenix quite bit is that the cluster configurations still have to match in several aspects, like timeouts, TLS/SASL settings, etc, otherwise the client either can not even connect, or experiences errors due timeout / buffer size etc mismathces.
> 
> I think that some of that may also be a problem when configuring replication.

For replication in hbase, there is configuration map in peer configuration, so we could add configurations specific for connecting to the peer cluster.